crates/wasi-common/src/ctx.rs

```diff
@@ -408,42 +410,52 @@ impl WasiCtx {
     }
 
     /// Check if `WasiCtx` contains the specified raw WASI `fd`.
-    pub(crate) unsafe fn contains_entry(&self, fd: wasi::__wasi_fd_t) -> bool {
-        self.entries.contains_key(&fd)
+    pub(crate) unsafe fn contains_entry(&self, fd: types::Fd) -> bool {
+        self.entries.borrow().contains_key(&fd)
     }
 
     /// Get an immutable `Entry` corresponding to the specified raw WASI `fd`.
-    pub(crate) unsafe fn get_entry(&self, fd: wasi::__wasi_fd_t) -> WasiResult<&Entry> {
-        self.entries.get(&fd).ok_or(WasiError::EBADF)
+    pub(crate) unsafe fn get_entry(&self, fd: types::Fd) -> Result<Ref<Entry>> {
+        if !self.contains_entry(fd) {
+            return Err(Errno::Badf);
+        }
+        Ok(Ref::map(self.entries.borrow(), |entries| {
+            entries.get(&fd).unwrap()
```

Comment:
Awesome, this is exactly what I wanted to do and couldn't figure out how. Thanks! :-D

Reply:
Ok, this proves to be more difficult than I originally envisioned. If I follow your suggestion, than we end up with a lifetime mismatch since the first `maybe` will contain weirdness of the like `&Option<&Entry>`, and the compiler cannot assign approriate lifetimes to the final `Ref<'_, Entry>`. Perhaps when  we revisit the use of `RefCell` in `wasi-common` altogether, this will no longer be an issue?

Reply:
Ok, if it gets awkward, then I agree it makes sense to leave the code as-is for now.